File: docs/test-cases/logging-failures.go
// Test case: logging-failures (A09:2025)
package main

import (
	"log"
	"net/http"
)

func apiCall(w http.ResponseWriter, r *http.Request) {
	token := r.Header.Get("Authorization")
	// BUG: bearer token leaked into logs
	log.Printf("token=%s", token)
	doWork(token)
}

func login(w http.ResponseWriter, r *http.Request) {
	user := r.FormValue("user")
	pass := r.FormValue("pass")
	if !authenticate(user, pass) {
		// BUG: failed authentication path emits no security log entry
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	w.Write([]byte("ok"))
}

func search(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query().Get("q")
	// BUG: user input passed as format string (Printf vuln) and unescaped CRLF log injection
	log.Printf(q)
	runQuery(q)
}

func deleteAccount(w http.ResponseWriter, r *http.Request) {
	id := r.URL.Query().Get("id")
	store.Delete(id)
	// BUG: destructive action performed with no audit log of actor/target/time
	w.Write([]byte("deleted"))
}

File: docs/test-cases/logging-failures.rs
// Test case: logging-failures (A09:2025)
use axum::{extract::Query, http::StatusCode};
use serde::Deserialize;

#[derive(Deserialize)]
struct LoginReq { user: String, pass: String }

#[derive(Deserialize)]
struct Search { q: String }

async fn call_upstream(key: String) {
    // BUG: API key written to structured log
    tracing::info!("apikey={}", key);
    upstream::send(&key).await;
}

async fn login(Query(req): Query<LoginReq>) -> StatusCode {
    if !auth::verify(&req.user, &req.pass) {
        // BUG: authentication failure path emits no log event
        return StatusCode::UNAUTHORIZED;
    }
    StatusCode::OK
}

async fn search(Query(s): Query<Search>) {
    // BUG: user-controlled string logged unescaped — CRLF/log injection
    log::info!("{}", s.q);
    db::query(&s.q).await;
}

async fn admin_purge(Query(t): Query<Search>) {
    admin::purge(&t.q).await;
    // BUG: privileged admin action logged with no actor identity
    tracing::info!("purge executed target={}", t.q);
}
